JsonStar.Serialization.Demo/JsonStar.Serialization.Demo3.fst

```diff
@@ -0,0 +1,198 @@
+        if is_record_type e typ_term
+            then begin
+                let deserializer_f = create_deserializer_f e t b in
+                exact deserializer_f
+                end
+            else fail "We only support deserialization of a record, sorry!"
+        end
+    | _ -> fail "Invalid usage of deserialize_impl_dyn. It should be used to generate a function: 'json -> result t'"
+
+
+class serializable a = {
+    //serialize_gen: a -> result json; 
+    deserialize_gen: (j:json{JObject? j}) -> result a
+}
+
+//let serialize   #a {| serializable a |} v = serialize_gen   #a v
+let deserialize #a {| serializable a |} v = deserialize_gen #a v
+
+/// Example record type that we would want to deserialize //
+type r =                                                  //
+    {                                                     //
+        l : x:int{x >= 5};                                //
+    }                                                     //
+////////////////////////////////////////////////////////////
+
+// NOTE1: 
+//     We can't inspect type at runtime with F*, so we need some boilerplate to 
+//     have a nice, generic serialization function. For each type we need to define an instance of 
+//     a type class.
+// NOTE2: 
+//     Unfortunately we can't simply pass json argument to 'deserialize_impl' and pattern match on it. 
+//     We actually need to produce a function type with Meta-F* that takes a json argument and does something with it. 
+instance r_serializable: serializable r = {
+   //serialize_gen = (fun x -> synth_by_tactic (fun () -> serialize_impl (`r) x));
+   deserialize_gen = synth_by_tactic (fun () -> deserialize_impl r (`r));
+}
+
+// We could simplify it a little and generate these instances with a line similar to the following:
+// %splice[] (generate_serializable (`r))
+// where generate_serializable is a Meta-F* function that realizes the type class instance. 
+// Still it's an extra line per type to make it json-serializable. 
+// The good part is that we should be able to add these instances on top of library types as long 
+// as their definition is visible/accessible. 
+
+    
+let process_r_from_json (j : json{JObject? j}) : Tot (result (x:int{x >= 5})) = 
+    let r_deserialized : result r = deserialize j in
+    match r_deserialized with
+    | Valid rv -> Valid rv.l
+    | Error msg -> Error msg
+
+let main = 
+    FStar.IO.print_string "Write json:\n";
+    let json_str = FStar.IO.input_line () in
+    let json = JsonStar.Parser.parse json_str in
+    if JObject? json
+        then begin
+            let l_r = process_r_from_json json in
+            match l_r with
+            | Valid l -> FStar.IO.print_string (P.sprintf "Value of l field is: %d\n" l)
+            | Error msg -> FStar.IO.print_string (P.sprintf "Value of l field is not available due to: %s\n" msg)
+            end
+        else FStar.IO.print_string "Provided json does not represent an object\n"
```

Comment:
Why do you have this restriction? What would you expect the serialization format for e.g. an `int` to be? Maybe the restriction belongs better in the actual deserializer.

Reply:
My goal was to simplify the example by requiring that deserialize_impl takes 'j:json{JObject? j}' instead of any json. For the real library that restriction would have to be removed. I agree that moving this restriction into deserialize_impl makes more sense, I'll do that.

Reply:
On a second thought, I'll leave it like it is for now. The extra match and error handling would have to happen in the generated function (as the json is only available at runtime). It's not hard, but it would make the code less readable.